File: chapter02/LessThan.rb
class LessThan < Struct.new(:left, :right)
  def reduce(environment)
    if left.reducible?
      LessThan.new(left.reduce(environment), right)
    elsif right.reducible?
      LessThan.new(left, right.reduce(environment))
    else
      Boolean.new(left.value < right.value)
    end
  end

  def reducible?
    false
  end

  def to_s
    "#{left} < #{right}"
  end

  def inspect
    "#{self}"
  end
end

File: SmallStepSemantic/LessThan.py
from chapter02.SmallStepSemantic.Boolean import Boolean


class LessThan(object):
    def __init__(self, left, right):
        self.left = left
        self.right = right

    @staticmethod
    def reducible():
        return True

    def reduce(self, environment):
        if self.left.reducible():
            return LessThan(self.left.reduce(environment), self.right)
        elif self.right.reducible():
            return LessThan(self.left, self.right.reduce(environment))
        else:
            return Boolean(self.left.value < self.right.value)

    def to_s(self):
        return self.left.to_s() + ' < ' + self.right.to_s()